src/vm/opcodes.rs

```diff
@@ -20,37 +20,288 @@
 // limitations under the License.
 
-#![allow(clippy::unusual_byte_groupings)]
+use std::marker::PhantomData;
 
-use aluvm::isa::opcodes::INSTR_ISAE_TO;
+use aluvm::reg::{Reg16, Reg32, RegS};
+use amplify::num::{u2, u3};
 
-pub const INSTR_RGBISA_FROM: u8 = 0b10_010_000;
-pub const INSTR_RGBISA_TO: u8 = INSTR_ISAE_TO;
+use super::ContractStateAccess;
 
-// CONTRACTS:
-pub const INSTR_CNP: u8 = 0b11_000_000;
-pub const INSTR_CNS: u8 = 0b11_000_001;
-pub const INSTR_CNG: u8 = 0b11_000_010;
-pub const INSTR_CNC: u8 = 0b11_000_011;
+/// Operations defined under RGB ISA extension (`RGB`).
+#[derive(Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Display)]
+pub enum ContractOp<S: ContractStateAccess> {
+    /// Counts number of global items elements defined by the current operation of the type,
+    /// provided by the second argument, and puts the number to the destination `a32`
+    /// register from the first argument.
+    ///
+    /// If the operation doesn't contain inputs with a given assignment type, sets destination
+    /// index to zero.
+    ///
+    /// # Idempotence
+    ///
+    /// The operation is idempotent.
+    ///
+    /// # Fails
+    ///
+    /// Operation fails by setting `st0` to fail state and terminating the program. This happens
+    /// only if the `ty` index is unset. In this case, the value of the destination register
+    /// remains unchanged.
+    ///
+    /// If operation doesn't fail, the value of `st0` remains unaffected (i.e. if it was set to
+    /// failed state before the operation, the operation doesn't change that).
+    #[display("cn.c    a32{dst},a16{ty}")]
+    CnC {
+        /// Index of an `a32` register receiving count of global state items of the type provided
+        /// in `ty`, contained in the contract global state.
+        dst: Reg32,
+        /// Index of `a16` register containing global state type.
+        ty: Reg32,
+    },
 
-pub const INSTR_LDP: u8 = 0b11_000_100;
-pub const INSTR_LDS: u8 = 0b11_000_101;
-// Reserved 0b11_000_111
+    /// Counts number of global items elements defined by the current operation of the type,
+    /// provided by the second argument, and puts the number to the destination `a16`
+    /// register from the first argument.
+    ///
+    /// If the operation doesn't contain inputs with a given assignment type, sets destination
+    /// index to zero.
+    ///
+    /// # Idempotence
+    ///
+    /// The operation is idempotent.
+    ///
+    /// # Fails
+    ///
+    /// Operation fails by setting `st0` to fail state and terminating the program. This happens
+    /// only if the `ty` index is unset. In this case, the value of the destination register
+    /// remains unchanged.
+    ///
+    /// If operation doesn't fail, the value of `st0` remains unaffected (i.e. if it was set to
+    /// failed state before the operation, the operation doesn't change that).
+    #[display("cn.g    a16{dst},a16{ty}")]
```

Comment:
Just a recap:

Reg32, Reg16 and Reg8 are just different bit-length for the registry index, it has nothing to do with a8/a16/a32, right?

Reply:
Yes, correct